hip/test/components/merging.hip.cpp

```diff
@@ -0,0 +1,294 @@
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+// TODO remove when the HIP includes are fixed
+#include <hip/hip_runtime.h>
+
+
+#include "hip/components/merging.hip.hpp"
+
+
+#include <algorithm>
+#include <memory>
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/executor.hpp>
+
+
+#include "hip/components/cooperative_groups.hip.hpp"
+
+
+namespace {
+
+
+using namespace gko::kernels::hip;
+using namespace gko::kernels::hip::group;
+
+
+class Merging : public ::testing::Test {
+protected:
+    Merging()
+        : ref(gko::ReferenceExecutor::create()),
+          hip(gko::HipExecutor::create(0, ref)),
+          rng(123456),
+          rng_runs{100},
+          max_size{1637},
+          sizes{0,  1,  2,   3,   4,   10,  15,   16,
+                31, 34, 102, 242, 534, 956, 1239, 1637},
+          data1(ref, max_size),
+          data2(ref, max_size),
+          outdata(ref, 2 * max_size),
+          idxs1(ref),
+          idxs2(ref),
+          idxs3(ref),
+          refidxs1(ref),
+          refidxs2(ref),
+          refidxs3(ref),
+          refdata(ref, 2 * max_size),
+          ddata1(hip),
+          ddata2(hip),
+          didxs1(hip, 2 * max_size),
+          didxs2(hip, 2 * max_size),
+          didxs3(hip, 2 * max_size),
+          drefidxs1(hip, 2 * max_size),
+          drefidxs2(hip, 2 * max_size),
+          drefidxs3(hip, 2 * max_size),
+          doutdata(hip, 2 * max_size)
+    {}
+
+    void init_data(int rng_run)
+    {
+        std::uniform_int_distribution<gko::int32> dist(0, max_size);
+        std::fill_n(data1.get_data(), max_size, 0);
+        std::fill_n(data2.get_data(), max_size, 0);
+        for (auto i = 0; i < max_size; ++i) {
```

Comment:
Please no `auto` in combination with an integer literal.
```suggestion
        for (int i = 0; i < max_size; ++i) {
```